### DISCUSSION TITLE
@keystone-6 serverless support

### DISCUSSION THREAD

**Discussion by jlarmstrongiv** [keystonejs/keystone#8066] [Feature requests]
KeystoneJS is surprisingly close to supporting serverless environments, like AWS Lambda.  With a concentrated push, we can knock an item off the [roadmap](https://keystonejs.com/updates/roadmap).

It is possible to use KeystoneJS with serverless today by following this outline:

1. Create a `getExpressServer` function modeled after the [start](https://github.com/keystonejs/keystone/blob/main/packages/core/src/scripts/run/start.ts) command
2. Recreate helpful utilities that `getExpressServer` depends on
    - [loadBuiltConfig](https://github.com/keystonejs/keystone/blob/4ec75f388db8157f757307b828354caabd3bb669/packages/core/src/lib/config/loadConfig.ts)
    - [requirePrismaClient](https://github.com/keystonejs/keystone/blob/main/packages/core/src/artifacts.ts)
    - [other helpful functions](https://github.com/keystonejs/keystone/blob/main/packages/core/src/scripts/utils.ts)
3. Create a bootstrap and handler function based off [NestJS](https://docs.nestjs.com/faq/serverless) using `@vendia/serverless-express`
4. Deploy with serverless using [container images](https://www.serverless.com/blog/container-support-for-lambda)
    - Pick a container image like `public.ecr.aws/lambda/nodejs:16`, or any other image where you add `aws-lambda-ric`
    - Use `tsup` with `noExternal` or `@vercel/ncc` to bundle your handler entrypoint into a single file for size, performance, and monorepo support
    - Exclude `@keystone-6/core` from the bundle or you will find this error https://github.com/keystonejs/keystone/issues/7816
    - Install `@keystone-6/core`, copy your `dist` folder and the `.prisma` and `.keystone` folders to your Docker image
    - Ensure that your prisma and other NodeJS C++ native addons are compiled for Linux, not MacOS

Also, use a database connection pooler like `pg-pool` in serverless environments.

---

For true serverless support, I would love to be able to bundle `@keystone-6/core` without running into graphql errors.

For experimental serverless support, I would like to see an experimental `getExpressServer` export, to avoid depending too heavily on KeystoneJS externals that may change unexpectedly during patches.  

If no serverless support will be added at this time, hopefully this guide will point others in the right direction.

I plan to link a demo github repo, but I’m sick this week.  If there’s interest, I will make one.

---

Followup from https://github.com/keystonejs/keystone/discussions/6806

**Comment by junaid33** [comment]
I was able to get the next-embedded example deployed serverlessly to lambda and vercel with the Admin UI and GraphQL API:

https://github.com/openshiporg/openfront

I used dynamic paths and getServiceSideProps to rebuild the Admin UI to be used for any Next.js frontend

**Reply by mariomnts** [reply]
This is cool! Where you able to deploy to AWS Lambda using a normal bundle or a docker image? Also, what are the performance of it? In my tests I saw cold Lambdas taking too much time to spin Prisma and connect to the DB

**Reply by junaid33** [reply]
I personally deployed this to Netlify and Vercel (both use Lambda under the hood), but they do some magic for cold starts. I mainly use serverful hosting like Render.com, but want the app ready for serverless in the future.

**Comment by kerbe** [comment]
I'm definitely interested seeing demo github repo @jlarmstrongiv ! I've been thinking of ways to get KeystoneJS running in AWS Lambda, but haven't got too deep in it.

I just got NextJS based application to run on AWS Lambda (not Lambda@Edge), with `serverless-http`. Now that I realized that KeystoneJS is build on top of NextJS as well, I have hope that it could be done. Tried to find way / place where to start in source code to start tinkering, but a bit complex... So demo would help a lot.

So, if there is some movement towards making this happen, I'm interested. And I'm willing to brainstorm ways to get things work, even though my technical skills might not be up to bar to implement them.

**Reply by jlarmstrongiv** [reply]
Hey @kerbe, see the followup discussion in [Slack](https://keystonejs.slack.com/archives/C01STDMEW3S/p1667518927673459).  

While the approach works, it depends on KeystoneJS internals.  You have the gist of it, using something like `serverless-http` or `@vendia/serverless-express` with the underlying KeystoneJS server.  Unfortunately:

> Would you be open to a PR that exposes a function like getExpressServer that returns the expressApp?

> Probably against that idea, for now.
> We will hopefully provide you with the tools that make that unnecessary.

So that rules out serverless and other hosting options without relying on KeystoneJS internals.  And why not rely on KeystoneJS internals?  Because they’re undocumented and change!  In fact, when I made my own prototype, they completely changed on me in the middle of it.

That leaves a few options:
- keep trying to convince the KeystoneJS team to expose `getExpressServer`
- make third-party npm package that does the dirty work of `getExpressServer` and keep it up to date as KeystoneJS changes
- just use my own workaround and hope it doesn’t break™

I was pretty happy with the solution I described above because it worked!  And it worked today!  API, Admin UI, everything.  And while there was some latency, it wasn’t terrible.  Anyway

**Reply by junaid33** [reply]
Take a look at https://github.com/openshiporg/openfront and try deploying that serverlessly. My colleague was able to deploy it to lambda without using express at all. It's based on the next-embedded example, which uses ApolloServerMicro instead of ApolloServerExpress in production. ApolloServerMicro is also better suited for serverless deployment.

**Reply by kerbe** [reply]
> Hey @kerbe, see the followup discussion in [Slack](https://keystonejs.slack.com/archives/C01STDMEW3S/p1667518927673459).

I don't have access to KeystoneJS, as it allows only keystonejs.com and thinkmill.com.au email addresses to register. Haven't found way to get invite link, and not sure how to contact slack admins, as register page describes that I should be doing.

> That leaves a few options:
>     * keep trying to convince the KeystoneJS team to expose `getExpressServer`
>     * make third-party npm package that does the dirty work of `getExpressServer` and keep it up to date as KeystoneJS changes
>     * just use my own workaround and hope it doesn’t break™
> I was pretty happy with the solution I described above because it worked! And it worked today! API, Admin UI, everything. And while there was some latency, it wasn’t terrible. Anyway

I wonder if that last "Anyway" was cut off. Anyway, that sounds quite complicated, especially if first point doesn't raise interest in KeystoneJS team. Making separate npm package could be "fun", but definitely would be way to keep oneself busy in the future.


> Take a look at https://github.com/openshiporg/openfront and try deploying that serverlessly. My colleague was able to deploy it to lambda without using express at all. It's based on the next-embedded example, which uses ApolloServerMicro instead of ApolloServerExpress in production. ApolloServerMicro is also better suited for serverless deployment.

I took a quick look on that, and noticed it was using Vercel. Didn't see that it could work in AWS Lambda without bigger customization, so didn't try it out yet. If there are any configuration / deployment notes, I'd love to see those, as it most likely cuts a lot of frustration out of my trial.

**Reply by junaid33** [reply]
[Openfront](https://github.com/openshiporg/openfront) is a Next.js application so you would have to use [Serverless Next.js](https://github.com/serverless-nextjs/serverless-next.js) to deploy to AWS Lambda.

**Reply by jlarmstrongiv** [reply]
Followup discussion https://github.com/keystonejs/keystone/discussions/8091 😄
